src/time-to-interactive/TTIBoomerang.ts

```diff
@@ -0,0 +1,331 @@
+import { isFCPSupported, isLCPSupported } from '../utils/common-utils';
+import { onFCP, onLCP, Metric } from 'web-vitals';
+
+const LONG_TASK = 'longtask';
+const FPS = 'fps';
+const NAVIGATION = 'navigation';
+const FCP = 'FCP';
+const LCP = 'LCP';
+
+/*
+This implements the TTI Boomerang algorithm with some modifications. 
+(Ref: https://akamai.github.io/boomerang/oss/BOOMR.plugins.Continuity.html)
+
+Steps to TTI: 
+1) Find visually ready timestamp (highest of domcontentLoadedEnd(if available), FCP(if available), LCP(if available))
+2) Starting from the Visually Ready timestamp, find a 500ms quiet window. 
+
+A quiet window has the following characteristics: 
+a) No Long Tasks 
+b) FPS is above 20
+
+Boomerang TTI's measurement interval is every 100ms. Each check is performed and bucketed into a 100ms interval. 
+
+3) TTI is recorded as Visually ready timestamp + time from visually ready to the start of the quiet window
+
+
+We do not record PageBusy metrics for now as there were some performance issues observed. 
+As such TTI can only be computed for now for browsers that support Long Tasks. 
+*/
+
+export class TTIBoomerang {
+    private fcpTime: number | null = null;
+    private lcpTime: number | null = null;
+    private domContentLoadedEventEnd: number | null = null;
+    private visuallyReadyTimestamp: number | null = null;
+    private ttiTracker: any = {};
+    private ttiResolved = false;
+
+    private fcpSupported = false;
+    private lcpSupported = false;
+    private fpsSupported = false; // Check for support for requestAnimationFrame
+
+    private COLLECTION_PERIOD = 100;
+    private REQUIRED_OK_INTERVALS = 5;
+    private CHECK_PERIOD = 1000; // Perform another check for TTI/Visually ready
+    private VISUALLY_READY_RESOLVE_TIMEOUT = 10000;
+    private TTI_RESOLVE_TIMEOUT = 10000;
+
+    private FPS_THRESHOLD = 20 / (1000 / this.COLLECTION_PERIOD);
+    private LONG_TASK_THRESHOLD = 0;
+
+    // Needs to return a Promise
+    computeTimeToInteractive(): Promise<number> {
+        this.initListeners();
+        return new Promise<number>((resolve, reject) => {
+            this.checkForVisualReady().then((result) => {
+                this.computeTTI(result).then((result) => {
+                    resolve(result);
+                });
+            });
+        });
+    }
+
+    checkForVisualReady(): Promise<number> {
```

Comment:
nit: By convention, this package uses async functions instead of promises. Make this an async function.

Async functions are just promises with syntactic sugar.